File: lib/services/dashboard_service.dart
import 'package:fluire/modelos/dashboard.dart';
import 'package:fluire/modelos/frequencia.dart';
import 'package:fluire/services/alunos_service.dart';
import 'package:fluire/services/api_client.dart';
import 'package:fluire/services/aulas_service.dart';
import 'package:fluire/services/frequencia_service.dart';

class DashboardService {
  final ApiClient _api = ApiClient.instance;
  final AlunosService _alunosService = AlunosService();
  final AulasService _aulasService = AulasService();
  final FrequenciaService _frequenciaService = FrequenciaService();

  Future<DashboardData> buscarDashboard() async {
    try {
      final response = await _api.get('/dashboard');
      if (response.statusCode == 200) {
        final jsonData = ApiClient.decodificarCorpo(response);
        if (jsonData is Map<String, dynamic>) {
          return DashboardData.fromJson(jsonData);
        }
        if (jsonData is Map) {
          return DashboardData.fromJson(Map<String, dynamic>.from(jsonData));
        }
      }
    } catch (_) {
      // fallback abaixo
    }

    return await _montarDashboardLocal();
  }

  String _converterDiaSemanaInt(int dia) {
    switch (dia) {
      case 1:
        return 'segunda-feira';
      case 2:
        return 'terça-feira';
      case 3:
        return 'quarta-feira';
      case 4:
        return 'quinta-feira';
      case 5:
        return 'sexta-feira';
      case 6:
        return 'sábado';
      case 7:
        return 'domingo';
      default:
        return 'segunda-feira';
    }
  }

  Future<DashboardData> _montarDashboardLocal() async {
    final alunos = await _alunosService.listar();
    final aulas = await _aulasService.listar();
    List<Frequencia> frequencias = [];

    try {
      frequencias = await _frequenciaService.listar();
    } catch (_) {}

    final hoje = DateTime.now();
    final diaSemanaTexto = _converterDiaSemanaInt(hoje.weekday);
    final aulasHoje = aulas.where((a) => a.diaSemana.toLowerCase().trim() == diaSemanaTexto).toList();

    final hojeStr =
        '${hoje.year}-${hoje.month.toString().padLeft(2, '0')}-${hoje.day.toString().padLeft(2, '0')}';
    final presentesHoje = frequencias.where((f) {
      return f.dataPresenca.startsWith(hojeStr) && f.presente == 1;
    }).length;

    final totalFreq = frequencias.length;
    final totalPresentes = frequencias.where((f) => f.presente == 1).length;
    final media = totalFreq == 0 ? 0 : ((totalPresentes / totalFreq) * 100).round();

    return DashboardData(
      totalAlunos: alunos.length,
      totalAulas: aulas.length,
      alunosPresentes: presentesHoje,
      aulasHoje: aulasHoje.length,
      emAndamento: aulasHoje.isNotEmpty ? 1 : 0,
      frequenciaMedia: media,
      weeklyFrequency: _frequenciaSemanalDeDados(frequencias),
      todayClasses: aulasHoje
          .map((a) => {
                'title': a.nome,
                'teacher': a.professorNome.isNotEmpty ? a.professorNome : 'Professor',
                'time': a.horario,
                'students': '${a.alunoIds.length} alunos',
                'status': 'ativa',
              })
          .toList(),
      alunosComFalta: frequencias.where((f) => f.presente == 0).length,
    );
  }

  List<Map<String, dynamic>> _frequenciaSemanalDeDados(List<Frequencia> frequencias) {
    const dias = ['Seg', 'Ter', 'Qua', 'Qui', 'Sex', 'Sáb', 'Dom'];
    final contagem = List.filled(7, 0);

    for (final f in frequencias) {
      if (f.presente != 1) continue;
      final dataStr = f.dataPresenca;
      if (dataStr.isEmpty) continue;
      try {
        final data = DateTime.parse(dataStr);
        final idx = data.weekday - 1;
        if (idx >= 0 && idx < 7) contagem[idx]++;
      } catch (_) {}
    }

    final max = contagem.isEmpty ? 0 : contagem.reduce((a, b) => a > b ? a : b);
    return List.generate(7, (i) {
      final valor = max == 0 ? 0.0 : (contagem[i] / max) * 80.0 + 20.0;
      return {'day': dias[i], 'value': valor};
    });
  }
}

File: src/services/dashboardService.js
import api from './api';
import * as alunosService from './alunosService';
import * as aulasService from './aulasService';
import * as frequenciaService from './frequenciaService';

const DIAS_SEMANA = {
  1: 'segunda-feira', 2: 'terça-feira', 3: 'quarta-feira',
  4: 'quinta-feira', 5: 'sexta-feira', 6: 'sábado', 7: 'domingo',
};

export async function buscarDashboard() {
  // Ignora o endpoint /dashboard do backend (conforme Regra de Ouro) e faz o cálculo 100% no front-end
  return await montarDashboardLocal();
}

function getInicioFimDaSemanaAtual() {
  const now = new Date();
  // Considerando a semana de Segunda a Domingo
  const dayOfWeek = now.getDay() === 0 ? 6 : now.getDay() - 1; // Seg=0, Dom=6
  const inicioSemana = new Date(now);
  inicioSemana.setDate(now.getDate() - dayOfWeek);
  inicioSemana.setHours(0, 0, 0, 0);

  const fimSemana = new Date(inicioSemana);
  fimSemana.setDate(inicioSemana.getDate() + 6);
  fimSemana.setHours(23, 59, 59, 999);

  return { inicioSemana, fimSemana };
}

async function montarDashboardLocal() {
  const [alunos, aulas] = await Promise.all([
    alunosService.listar().catch(() => []),
    aulasService.listar().catch(() => []),
  ]);

  let frequencias = [];
  try {
    frequencias = await frequenciaService.listar();
  } catch { /* ignore */ }

  const hoje = new Date();
  const diaSemanaTexto = DIAS_SEMANA[hoje.getDay() === 0 ? 7 : hoje.getDay()];
  const aulasHoje = aulas.filter((a) =>
    a.diaSemana.toLowerCase().trim() === diaSemanaTexto
  );

  const hojeStr = `${hoje.getFullYear()}-${String(hoje.getMonth() + 1).padStart(2, '0')}-${String(hoje.getDate()).padStart(2, '0')}`;
  const presentesHoje = frequencias.filter((f) =>
    f.dataPresenca.startsWith(hojeStr) && f.presente === 1
  ).length;

  // Lógica da Média Semanal Real (apenas da semana atual)
  const { inicioSemana, fimSemana } = getInicioFimDaSemanaAtual();
  
  const frequenciasDaSemana = frequencias.filter((f) => {
    if (!f.dataPresenca) return false;
    const d = new Date(f.dataPresenca + 'T00:00:00');
    return d >= inicioSemana && d <= fimSemana;
  });

  const totalMarcacoesNaSemana = frequenciasDaSemana.length;
  const presencasNaSemana = frequenciasDaSemana.filter((f) => f.presente === 1).length;
  const faltasNaSemana = frequenciasDaSemana.filter((f) => f.presente === 0).length;
  const frequenciaMedia = totalMarcacoesNaSemana === 0 ? 0 : Math.round((presencasNaSemana / totalMarcacoesNaSemana) * 100);

  return {
    totalAlunos: alunos.length,
    totalAulas: aulas.length,
    alunosPresentes: presentesHoje,
    aulasHoje: aulasHoje.length,
    emAndamento: 0, // Será calculado precisamente no DashboardPage.jsx
    frequenciaMedia: frequenciaMedia,
    weeklyFrequency: frequenciaSemanalDeAulas(aulas),
    todayClasses: aulasHoje.map((a) => ({
      id: a.id,
      title: a.nome,
      teacher: a.professorNome || 'Professor',
      time: `${a.horarioInicio} - ${a.horarioFim}`,
      horarioInicio: a.horarioInicio,
      horarioFim: a.horarioFim,
      alunoIds: a.alunoIds || [],
      students: `${(a.alunoIds || []).length} alunos`,
      status: 'ativa',
    })),
    alunosComFalta: faltasNaSemana,
  };
}

function frequenciaSemanalDeAulas(aulas) {
  const diasNomes = ['segunda-feira', 'terça-feira', 'quarta-feira', 'quinta-feira', 'sexta-feira', 'sábado', 'domingo'];
  const diasSiglas = ['Seg', 'Ter', 'Qua', 'Qui', 'Sex', 'Sáb', 'Dom'];
  const contagemSemana = [0, 0, 0, 0, 0, 0, 0];

  for (const aula of aulas) {
    const diaNorm = (aula.diaSemana || '').toLowerCase().trim();
    const idx = diasNomes.indexOf(diaNorm);
    if (idx !== -1) {
      contagemSemana[idx]++;
    }
  }

  const max = Math.max(...contagemSemana, 0);
  return diasSiglas.map((day, i) => ({
    day,
    value: max === 0 ? 0 : (contagemSemana[i] / max) * 80 + 20, // Altura visual da barra (20% a 100%)
    count: contagemSemana[i] // Valor matemático real da quantidade de aulas no dia
  }));
}
